### PR TITLE
Migrate scatter and scatter_ from TH to ATen (non-TensorIterator version).

### PR DESCRIPTION
**Pull request by v0dro**
Fix for https://github.com/pytorch/pytorch/issues/24621

### CODE REVIEW

**Comment by v0dro** [comment]
@VitalyFedyunin @ezyang I'm a little confused about the test failure and CI setup here. The only failure is a wrong value for these tests:
```
Mar 20 06:41:53 ======================================================================
Mar 20 06:41:53 FAIL [0.049s]: test_gumbel_softmax_cuda_float16 (__main__.TestNNDeviceTypeCUDA)
Mar 20 06:41:53 ----------------------------------------------------------------------
Mar 20 06:41:53 Traceback (most recent call last):
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 224, in instantiated_test
Mar 20 06:41:53     result = test(self, device_arg, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9677, in test_gumbel_softmax
Mar 20 06:41:53     self._test_gumbel_softmax_straight_through(device, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9650, in _test_gumbel_softmax_straight_through
Mar 20 06:41:53     self.assertLess(z.abs().max().item(), 2.58)
Mar 20 06:41:53 AssertionError: 16.515625 not less than 2.58
Mar 20 06:41:53 
Mar 20 06:41:53 ======================================================================
Mar 20 06:41:53 FAIL [0.049s]: test_gumbel_softmax_cuda_float32 (__main__.TestNNDeviceTypeCUDA)
Mar 20 06:41:53 ----------------------------------------------------------------------
Mar 20 06:41:53 Traceback (most recent call last):
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 224, in instantiated_test
Mar 20 06:41:53     result = test(self, device_arg, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9677, in test_gumbel_softmax
Mar 20 06:41:53     self._test_gumbel_softmax_straight_through(device, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9650, in _test_gumbel_softmax_straight_through
Mar 20 06:41:53     self.assertLess(z.abs().max().item(), 2.58)
Mar 20 06:41:53 AssertionError: 16.513498306274414 not less than 2.58
Mar 20 06:41:53 
Mar 20 06:41:53 ======================================================================
Mar 20 06:41:53 FAIL [0.049s]: test_gumbel_softmax_cuda_float64 (__main__.TestNNDeviceTypeCUDA)
Mar 20 06:41:53 ----------------------------------------------------------------------
Mar 20 06:41:53 Traceback (most recent call last):
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 224, in instantiated_test
Mar 20 06:41:53     result = test(self, device_arg, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9677, in test_gumbel_softmax
Mar 20 06:41:53     self._test_gumbel_softmax_straight_through(device, dtype)
Mar 20 06:41:53   File "test_nn.py", line 9650, in _test_gumbel_softmax_straight_through
Mar 20 06:41:53     self.assertLess(z.abs().max().item(), 2.58)
Mar 20 06:41:53 AssertionError: 16.513498170970525 not less than 2.58
Mar 20 06:41:53 
Mar 20 06:41:53 ======================================================================
Mar 20 06:41:53 FAIL [0.028s]: test_one_hot_cuda (__main__.TestNNDeviceTypeCUDA)
Mar 20 06:41:53 ----------------------------------------------------------------------
Mar 20 06:41:53 Traceback (most recent call last):
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 685, in wrapper
Mar 20 06:41:53     method(*args, **kwargs)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 207, in instantiated_test
Mar 20 06:41:53     return test(self, device_arg)
Mar 20 06:41:53   File "test_nn.py", line 9256, in test_one_hot
Mar 20 06:41:53     self.assertEqual(t, expected)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 882, in assertEqual
Mar 20 06:41:53     assertTensorsEqual(x, y)
Mar 20 06:41:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 848, in assertTensorsEqual
Mar 20 06:41:53     self.assertLessEqual(max_err, prec, message)
Mar 20 06:41:53 AssertionError: tensor(1, device='cuda:0') not less than or equal to 1e-05 : 
Mar 20 06:41:53 
Mar 20 06:41:53 ----------------------------------------------------------------------
Mar 20 06:41:53 Ran 1662 tests in 886.231s
Mar 20 06:41:53 
Mar 20 06:41:53 FAILED (failures=4, skipped=79, expected failures=4)
```

Since I'm not able to reproduce this on my machine I tried to log into the circle CI and lauch the docker image to see if I can run those tests on the build machine itself. However that fails with the following errors:
```
circleci@ip-172-16-138-86:~$ docker images
REPOSITORY                                                                                           TAG                                                                             IMAGE ID            CREATED             SIZE
308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10.2-cudnn7-py3-gcc7   f990c76a-a798-42bb-852f-5be5006f8026-c54d94c537bcd6f8dfe1633c62609d0a38796bf1   b2175d486b0d        3 hours ago         11.2GB
circleci@ip-172-16-138-86:~$ docker run -it 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10.2-cudnn7-py3-gcc7 bash
Unable to find image '308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10.2-cudnn7-py3-gcc7:latest' locally
docker: Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10.2-cudnn7-py3-gcc7:latest not found.
See 'docker run --help'.
circleci@ip-172-16-138-86:~$ docker run -it f990c76a-a798-42bb-852f-5be5006f8026-c54d94c537bcd6f8dfe1633c62609d0a38796bf1 bash
Unable to find image 'f990c76a-a798-42bb-852f-5be5006f8026-c54d94c537bcd6f8dfe1633c62609d0a38796bf1:latest' locally
docker: Error response from daemon: pull access denied for f990c76a-a798-42bb-852f-5be5006f8026-c54d94c537bcd6f8dfe1633c62609d0a38796bf1, repository does not exist or may require 'docker login'.
See 'docker run --help'.
```

Could you please point out whats going wrong?

**Comment by v0dro** [comment]
Benchmarks for the new implementation:
Script:
``` python
import torch
import numpy
from IPython import get_ipython

numpy.random.seed(13)
torch.manual_seed(13)

ipython = get_ipython()

Ms=256
Ns=512
dim = 0
top_power = 5

for pM in range(top_power):
    M = Ms * (2 ** pM)
    for pN in range(top_power):
        N = Ns * (2 ** pN)
        input_ = torch.rand(M, N, device=torch.device('cuda'))
        src = torch.rand(M, N, device=torch.device('cuda'))
        index = torch.tensor(numpy.random.randint(0, M, (M, N)), device=torch.device('cuda') )
        print(f"Problem size (MxN): {M}x{N}")
        ipython.magic("timeit input_.scatter_(dim, index, src)")
```

Output:
```
Problem size (MxN): 256x512
7.87 µs ± 23 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 256x1024
10.7 µs ± 27 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 256x2048
21.1 µs ± 48 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x4096
52.2 µs ± 32.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x8192
490 µs ± 12.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x512
11.1 µs ± 32 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 512x1024
22.1 µs ± 53.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x2048
52.7 µs ± 32.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x4096
491 µs ± 10.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x8192
1.18 ms ± 13.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x512
22.3 µs ± 63.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x1024
53.2 µs ± 37 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x2048
492 µs ± 17.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x4096
1.23 ms ± 18.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x8192
2.68 ms ± 50.6 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x512
53.5 µs ± 114 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x1024
455 µs ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x2048
1.23 ms ± 26.2 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x4096
2.68 ms ± 103 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x8192
5.72 ms ± 98.1 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x512
493 µs ± 6.94 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 4096x1024
1.23 ms ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 4096x2048
2.69 ms ± 112 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x4096
5.72 ms ± 159 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x8192
11.8 ms ± 104 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by nikitaved** [comment]
How does it compare to the TH version?

**Comment by v0dro** [comment]
@nikitaved I am running those benchmarks now and will post them here soon.

File: test/test_nn.py [hunk 1]
```diff
@@ -9420,9 +9420,14 @@ def _test_gumbel_softmax_straight_through(self, device, dtype):
         counts = torch.zeros_like(logits)
         for _ in range(num_draws):
             y_draw = F.gumbel_softmax(logits, hard=True)
+            print(y_draw)
             counts = counts + y_draw
 
         # All values positive
+        print("y_draw::")
```
Comment (VitalyFedyunin) [inline test/test_nn.py]: Please remove print when it is ready for review.

**Comment by v0dro** [comment]
@VitalyFedyunin request review.

**Comment by v0dro** [comment]
**Benchmarks**

Script:
``` python
import torch
import numpy
from IPython import get_ipython

numpy.random.seed(13)
torch.manual_seed(13)

ipython = get_ipython()

Ms=256
Ns=512
dim = 0
top_power = 5

for pM in range(top_power):
    M = Ms * (2 ** pM)
    for pN in range(top_power):
        N = Ns * (2 ** pN)
        input_ = torch.rand(M, N, device=torch.device('cuda'))
        src = torch.rand(M, N, device=torch.device('cuda'))
        index = torch.tensor(numpy.random.randint(0, M, (M, N)), device=torch.device('cuda') )
        print(f"Problem size (MxN): {M}x{N}")
        ipython.magic("timeit input_.scatter_(dim, index, src)")
```

Old implementation (current master):
```

```

New implementation:
```
Problem size (MxN): 256x512
6.64 µs ± 2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 256x1024
10.6 µs ± 23.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 256x2048
21.1 µs ± 54.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x4096
60.6 µs ± 13.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x8192
496 µs ± 10 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x512
11 µs ± 13.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
Problem size (MxN): 512x1024
22.1 µs ± 8.75 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x2048
52.7 µs ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x4096
497 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x8192
1.22 ms ± 10.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x512
22.2 µs ± 69.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x1024
53 µs ± 28.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x2048
497 µs ± 9.84 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x4096
1.27 ms ± 15.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x8192
2.78 ms ± 95.4 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x512
53.4 µs ± 102 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x1024
461 µs ± 7.57 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x2048
1.27 ms ± 17.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x4096
2.78 ms ± 60.8 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x8192
5.94 ms ± 79.7 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x512
499 µs ± 8.52 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 4096x1024
1.22 ms ± 12.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 4096x2048
2.78 ms ± 84.4 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x4096
5.95 ms ± 108 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x8192
12.3 ms ± 145 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by v0dro** [comment]
CC: @nikitaved

**Comment by VitalyFedyunin** [comment]
On review. Meanwhile please fix benchmarking script, as you need to call `torch.cuda.syncronize()` after each _scatter call.

**Comment by nikitaved** [comment]
> On review. Meanwhile please fix benchmarking script, as you need to call `torch.cuda.syncronize()` after each _scatter call.

After each call? Would after each magic call be sufficient?

**Comment by v0dro** [comment]
New script:
``` python
import torch
import numpy
from IPython import get_ipython

numpy.random.seed(13)
torch.manual_seed(13)

ipython = get_ipython()

Ms=256
Ns=512
dim = 0
top_power = 5

for pM in range(top_power):
    M = Ms * (2 ** pM)
    for pN in range(top_power):
        N = Ns * (2 ** pN)
        input_ = torch.rand(M, N, device=torch.device('cuda'))
        src = torch.rand(M, N, device=torch.device('cuda'))
        index = torch.tensor(numpy.random.randint(0, M, (M, N)), device=torch.device('cuda') )
        print(f"Problem size (MxN): {M}x{N}")
        ipython.magic("timeit input_.scatter_(dim, index, src); torch.cuda.synchronize()")
```
Master branch:
```
Problem size (MxN): 256x512
19.7 µs ± 650 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x1024
22.2 µs ± 129 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x2048
32.1 µs ± 131 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x4096
70.8 µs ± 360 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 256x8192
466 µs ± 553 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x512
23 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x1024
33 µs ± 272 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x2048
61.6 µs ± 57.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 512x4096
501 µs ± 690 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 512x8192
1.24 ms ± 248 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 1024x512
33 µs ± 95.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x1024
62 µs ± 24.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 1024x2048
503 µs ± 166 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 1024x4096
1.19 ms ± 616 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 1024x8192
2.75 ms ± 392 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 2048x512
62.8 µs ± 291 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Problem size (MxN): 2048x1024
506 µs ± 467 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x2048
1.24 ms ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 2048x4096
2.76 ms ± 843 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 2048x8192
5.74 ms ± 798 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 4096x512
504 µs ± 371 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x1024
1.19 ms ± 362 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Problem size (MxN): 4096x2048
2.75 ms ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 4096x4096
5.73 ms ± 3.32 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Problem size (MxN): 4096x8192
11.8 ms ± 598 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

New porting:
```
```

**Comment by v0dro** [comment]
New benchmarks:

# Script 1

``` python
import torch
from IPython import get_ipython

ipython = get_ipython()

torch.set_num_threads(1)

device = 'cuda'

nfeat = 10000
nrep = 5
a=torch.arange(nfeat, device=device).repeat_interleave(nrep)
batch=3 #batch can vary 1-200

res = torch.randn(100000, 100000, device=device)


for batch in [100, 500, 1000, 5000, 10000]:
    print("Batch: ", batch)
    c=torch.randint(3, (batch, nfeat * nrep), device=device).float()
    ipython.magic("timeit res.scatter_(1,a.unsqueeze(0).expand(batch,a.size(0)),c); torch.cuda.synchronize()")

    enum_values = [
        list(range(1, 201)),
        list(range(1000, 1020)),
        list(range(2000, 2010)),
        list(range(3000, 3206)),
    ]
    indices = torch.tensor([i for i, values in enumerate(enum_values) for _j in range(len(values))], device=device)
    c = torch.randint(3, (batch, 4), device=device).float()
    idx = indices.unsqueeze(0).expand(c.size(0), indices.size(0))
    src = c.repeat(1, idx.shape[-1] // c.shape[-1])
    ipython.magic("timeit res.scatter_(1,idx,src); torch.cuda.synchronize()")
    print()
```

## ATen

``` python
Batch:  100
112 µs ± 461 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
13 µs ± 242 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  500
485 µs ± 934 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
14.8 µs ± 19.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  1000
964 µs ± 1.51 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
18.9 µs ± 37.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  5000
4.84 ms ± 9.25 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
55 µs ± 125 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Batch:  10000
9.78 ms ± 5.52 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
95.1 µs ± 205 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

## TH

# Script 2

``` python
import torch
from IPython import get_ipython

ipython = get_ipython()

torch.set_num_threads(1)

device = 'cuda'

nrows = 100000
ncols = 100000
dims = [nrows, ncols]

res = torch.randn(dims, device=device)
idx1 = torch.randint(dims[0], (1, dims[1]), device=device).long()
src1 = torch.randn(1, dims[1], device=device)
idx2 = torch.randint(dims[1], (dims[0], 1), device=device).long()
src2 = torch.randn(dims[0], 1, device=device)

ipython.magic("timeit res.scatter_(0, idx1, src1); torch.cuda.synchronize()")
ipython.magic("timeit res.scatter_(1, idx2, src2); torch.cuda.synchronize()")
```

## ATen

``` python
198 µs ± 949 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
42.3 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
## TH

# Script 3

``` python
import torch
from IPython import get_ipython

ipython = get_ipython()

torch.set_num_threads(1)

device = 'cuda'

nfeat = 10000
nrep = 5
a=torch.arange(nfeat, device=device).repeat_interleave(nrep)
batch=3 #batch can vary 1-200

res = torch.randn(100000, 100000, device=device)


for batch in [100, 500, 1000, 5000, 10000]:
    print("Batch: ", batch)
    c=torch.randint(3, (batch, nfeat * nrep), device=device).float()
    ipython.magic("timeit res.scatter_(1,a.unsqueeze(0).expand(batch,a.size(0)),c); torch.cuda.synchronize()")

    enum_values = [
        list(range(1, 201)),
        list(range(1000, 1020)),
        list(range(2000, 2010)),
        list(range(3000, 3206)),
    ]
    indices = torch.tensor([i for i, values in enumerate(enum_values) for _j in range(len(values))], device=device)
    c = torch.randint(3, (batch, 4), device=device).float()
    idx = indices.unsqueeze(0).expand(c.size(0), indices.size(0))
    src = c.repeat(1, idx.shape[-1] // c.shape[-1])
    ipython.magic("timeit res.scatter_(1,idx,src); torch.cuda.synchronize()")
    print()
```
## ATen

``` python
Batch:  100
114 µs ± 406 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
13.1 µs ± 395 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  500
492 µs ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
14.7 µs ± 33.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  1000
969 µs ± 1.08 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
18.8 µs ± 52.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

Batch:  5000
4.85 ms ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
54.8 µs ± 125 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Batch:  10000
9.78 ms ± 7.53 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
94.9 µs ± 86.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

## TH

``` python
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1500,6 +1500,8 @@ def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
     if hard:
         # Straight through.
         index = y_soft.max(dim, keepdim=True)[1]
+        print("index:")
```
Comment (VitalyFedyunin) [inline torch/nn/functional.py:1503]: please remove print statements

File: test/test_nn.py [hunk 3]
```diff
@@ -9420,6 +9420,7 @@ def _test_gumbel_softmax_straight_through(self, device, dtype):
         counts = torch.zeros_like(logits)
         for _ in range(num_draws):
             y_draw = F.gumbel_softmax(logits, hard=True)
+            print(y_draw)
```
Comment (VitalyFedyunin) [inline test/test_nn.py:9423]: Please remove

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 4]
```diff
@@ -298,92 +198,4 @@ void THCTensor_(scatterAdd)(THCState* state, THCTensor *tensor, int dim, THCudaL
...
-
-void
-THCTensor_(scatterFill)(THCState* state, THCTensor *tensor,
-                         int dim, THCudaLongTensor *index, scalar_t value) {
-  THCAssertSameGPU(THCTensor_(checkGPU)(state, 1, tensor));
-  THCAssertSameGPU(THCudaLongTensor_checkGPU(state, 1, index));
-
-  dim = at::maybe_wrap_dim(dim, tensor);
-  int index_ndim_legacy_all = THCudaLongTensor_nDimensionLegacyAll(state, index);
-  THArgCheck(dim >= 0 && dim < THCTensor_(nDimensionLegacyNoScalars)(state, tensor), 2,
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:314]: Where did this check go?

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 5]
```diff
@@ -298,92 +198,4 @@ void THCTensor_(scatterAdd)(THCState* state, THCTensor *tensor, int dim, THCudaL
...
-             "Index dimension is out of bounds");
-  THArgCheck(index_ndim_legacy_all == 0
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:316]: Where did this check go?

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 6]
```diff
@@ -298,92 +198,4 @@ void THCTensor_(scatterAdd)(THCState* state, THCTensor *tensor, int dim, THCudaL
...
-             || THCudaLongTensor_nDimensionLegacyNoScalars(state, index) == THCTensor_(nDimensionLegacyNoScalars)(state, tensor), 3,
-             "Index tensor must be either empty or have same dimensions as output tensor");
-
-  // no-op if index is empty
-  if (index_ndim_legacy_all == 0)
-      return;
-
-  for (int d = 0; d < THCTensor_(nDimensionLegacyNoScalars)(state, tensor); d++) {
-    if (d != dim) {
-      THArgCheck(THCTensor_(sizeLegacyNoScalars)(state, tensor, d) ==
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:326]: New code missing this check

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 7]
```diff
@@ -98,106 +98,6 @@ void THCTensor_(gather)(THCState* state, THCTensor *tensor,
...
-  <<<grid, block, 0, c10::cuda::getCurrentCUDAStream(curDevice)>>>(               \
-    tensorInfo, srcInfo, indexInfo, dim, (TYPE)totalElements);
-
-void THCTensor_(scatter)(THCState* state, THCTensor *tensor, int dim, THCudaLongTensor *index, THCTensor *src) {
-  THCAssertSameGPU(THCTensor_(checkGPU)(state, 2, tensor, src));
-  THCAssertSameGPU(THCudaLongTensor_checkGPU(state, 1, index));
-
-  dim = at::maybe_wrap_dim(dim, tensor);
-  int index_ndim_legacy_all = THCudaLongTensor_nDimensionLegacyAll(state, index);
-  THArgCheck(dim >= 0 && dim < THCTensor_(nDimensionLegacyNoScalars)(state, tensor), 2,
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:113]: New code missing this check

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 8]
```diff
@@ -98,106 +98,6 @@ void THCTensor_(gather)(THCState* state, THCTensor *tensor,
...
-             "Index dimension is out of bounds");
-  THArgCheck(index_ndim_legacy_all == 0
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:115]: New code missing this check

**Comment by nikitaved** [comment]
@VitalyFedyunin, the TensorIterator port has been already merged.

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 9]
```diff
@@ -98,106 +98,6 @@ void THCTensor_(gather)(THCState* state, THCTensor *tensor,
...
-             "Input tensor must have same dimensions as output tensor");
-
-  // no-op if index is empty
-  if (index_ndim_legacy_all == 0)
-      return;
-
-  for (int d = 0; d < THCTensor_(nDimensionLegacyNoScalars)(state, tensor); d++) {
-    int64_t indexSizeD = THCudaLongTensor_sizeLegacyNoScalars(state, index, d);
-    if (d != dim) {
-      THArgCheck(indexSizeD <= THCTensor_(sizeLegacyNoScalars)(state, tensor, d), 3,
```
Comment (VitalyFedyunin) [inline aten/src/THC/generic/THCTensorScatterGather.cu:128]: New code missing this check

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Debug lines.
Lots of checks omitted. Could you please add tests that will trigger these conditions?

**Comment by rgommers** [comment]
Indeed, looks like this should have been closed 3 days ago when gh-35697 was merged - so let's close it now. Thanks for the PR @v0dro, and thanks for the review effort @VitalyFedyunin.

**Pull request closed without merging**